src/replication/stream_entry_fetcher.cpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * stream_entry_fetcher.cpp - Fetcher of stream entries from a stream
+ */
+
+#include "stream_entry_fetcher.hpp"
+#include "string_buffer.hpp"
+#include "thread_looper.hpp"
+
+namespace cubstream
+{
+  class stream_entry_fetch_task : public cubthread::entry_task
+  {
+    public:
+      stream_entry_fetch_task (stream_entry_fetcher &fetcher)
+	: m_fetcher (fetcher)
+      {
+      };
+
+      void execute (cubthread::entry &thread_ref) override
+      {
+	stream_entry *se = NULL;
+
+	int err = m_fetcher.fetch_stream_entry (se);
```

Comment:
This code may be included in a single method of fetcher 
fetch_stream_entry + m_on_fetch  + push_entry = produce
The new metthod produce may be protected with stream_entry_fetch_task class fried of stream_entry_fetcher.
the other may be private